cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -0,0 +1,573 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "indexing/construction/detail/utilities.cuh"
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/spatial_join.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/gather.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/reduce.h>
+
+#include <vector>
+
+namespace cuspatial {
+namespace detail {
+namespace {
+
+static uint8_t const max_warps_per_block  = 32;
```

Comment:
If threads_per_block is 256, how can max_warps_per_block be 32? Shouldn't this be 8?

Reply:
cc: @zhangjianting

Reply:
> cc: @zhangjianting
max_warps_per_block suggests the maximum warps per block. For CC>=3.0, the number is 1024,

Reply:
OK, but it is only used to size shared memory. Instead could use dynamically sized shared memory based on the *actual* warps per block.